// svelte-legos/dist/actions/confettiAction/index.js
import { eventListenerStore } from "../../stores/eventListenerStore";
import confetti from "canvas-confetti";
const defaultConfettiParams = {
    type: "simple",
    particleCount: 100,
    spread: 70,
    origin: { y: 0.5, x: 0.5 },
};
export function confettiAction(node, params = defaultConfettiParams) {
    let stop;
    const destroy = () => {
        stop && stop();
    };
    const update = (params) => {
        destroy();
        async function handleClick() {
            if (params.type === "simple") {
                confetti(params);
            }
            else {
                const end = Date.now() + 1 * 1000;
                const colors = ["#bb0000", "#ffffff"];
                (function frame() {
                    confetti({
                        particleCount: 2,
                        angle: 60,
                        spread: 55,
                        origin: { x: 0 },
                        colors: colors,
                    });
                    confetti({
                        particleCount: 2,
                        angle: 120,
                        spread: 55,
                        origin: { x: 1 },
                        colors: colors,
                    });
                    if (Date.now() < end) {
                        requestAnimationFrame(frame);
                    }
                })();
            }
        }
        ({ stop } = eventListenerStore("click", handleClick, node));
    };
    update({ ...defaultConfettiParams, ...params });
    return {
        update,
        destroy,
    };
}
